subgraphs/uniswap-forks/protocols/trader-joe/src/common/handlers/rewarderRate/rewardRateCalculator.ts

```diff
@@ -0,0 +1,292 @@
+import {
+  Address,
+  BigInt,
+  BigDecimal,
+  ethereum,
+  log,
+} from "@graphprotocol/graph-ts";
+import {
+  _MasterChef,
+  _MasterChefRewarder,
+  _MasterChefStakingPool,
+  _RewarderProbe,
+  LiquidityPool,
+} from "../../../../../../generated/schema";
+import { Rewarder } from "../../../../../../generated/templates/Rewarder/Rewarder";
+import {
+  BIGINT_ZERO,
+  BIGINT_ONE,
+  BIGINT_TEN,
+  BIGINT_HUNDRED,
+  ZERO_ADDRESS,
+  BIGDECIMAL_TEN,
+  BIGDECIMAL_HUNDRED,
+} from "../../../../../../src/common/constants";
+import { MasterChefV2TraderJoe } from "../../../../../../generated/MasterChefV2/MasterChefV2TraderJoe";
+
+// RewarderCalculator can be used to estimate the rewardRate of masterchef rewarders.
+// It defaults to getting the rate from the contract if available, and if not it will estimate it from
+// the increase in accrued rewards for some randomly chose accounts.
+export class RewarderCalculator {
+  private mc: _MasterChef;
+  private rewarder: _MasterChefRewarder;
+  private contract: Rewarder;
+
+  constructor(mc: _MasterChef, rewarder: _MasterChefRewarder) {
+    this.mc = mc;
+    this.rewarder = rewarder;
+    this.contract = Rewarder.bind(Address.fromString(rewarder.id));
+  }
+
+  public static attemptToRetrieveRewardRate(contract: Rewarder): BigInt | null {
+    let tokenPerSecCall = contract.try_tokenPerSec();
+    if (tokenPerSecCall.reverted) {
+      return null;
+    }
+
+    return tokenPerSecCall.value;
+  }
+
+  // calculateRewarderRate will happen once every 6 hours.
+  // There are 2 ways of calculating the reward rate:
+  //   1. If the rewarder contract exposes `tokenPerSec`, we'll use that.
```

Comment:
Do you know if the rewarder ever exposes a tokenPerBlock() that you could utilize?

Reply:
I haven't seen any but they have an example in their reference rewarder implementations. So they might appear at some point. I wanted to leave it out of the PR for now because translating `perBlock` to `perSecond` would involve a few more changes. 

I can add it after, and is a change that should not require a full resync, grafting should do.

Reply:
Okay sounds fine to me.